src/utils.rs

```diff
@@ -1,15 +1,80 @@
 //! Helper routines.
 
-use std::ops::Range;
-
+use anyhow::{Error, anyhow};
 use camino::Utf8Path;
 use github_actions_models::common::{
     Env,
     expr::{ExplicitExpr, LoE},
 };
+use jsonschema::{
+    BasicOutput::{Invalid, Valid},
+    Validator,
+    output::{ErrorDescription, OutputUnit},
+    validator_for,
+};
+use std::{collections::VecDeque, ops::Range};
+use std::{fmt::Write, sync::LazyLock};
 
 use crate::{audit::AuditInput, models::AsDocument};
 
+static ACTION_VALIDATOR: LazyLock<Validator> = LazyLock::new(|| {
+    validator_for(&serde_json::from_str(include_str!("./data/github-action.json")).unwrap())
+        .unwrap()
+});
+
+static WORKFLOW_VALIDATOR: LazyLock<Validator> = LazyLock::new(|| {
+    validator_for(&serde_json::from_str(include_str!("./data/github-workflow.json")).unwrap())
+        .unwrap()
+});
+
+pub(crate) fn validate_action(contents: String) -> Option<Error> {
+    match serde_yaml::from_str(&contents) {
+        Ok(workflow) => match ACTION_VALIDATOR.apply(&workflow).basic() {
+            Valid(_) => None,
+            Invalid(errors) => Some(parse_validation_errors(errors)),
+        },
+        Err(e) => Some(anyhow!("invalid YAML in action definition: {e}")),
+    }
+}
+
+pub(crate) fn validate_workflow(contents: String) -> Option<Error> {
+    match serde_yaml::from_str(&contents) {
+        Ok(workflow) => match WORKFLOW_VALIDATOR.apply(&workflow).basic() {
+            Valid(_) => None,
+            Invalid(errors) => Some(parse_validation_errors(errors)),
+        },
+        Err(_) => Some(anyhow!("invalid YAML in workflow definition")),
+    }
+}
+
+fn parse_validation_errors(errors: VecDeque<OutputUnit<ErrorDescription>>) -> Error {
+    let mut message = String::new();
+
+    for error in errors {
+        let description = error.error_description().to_string();
+        // HACK: error descriptions are sometimes a long rats' nest
+        // of JSON objects. We should render this in a palatable way
+        // but doing so is nontrivial, so we just skip them for now.
+        // NOTE: Experimentally, this seems to mostly happen when
+        // the error for an unmatched "oneOf", so these errors are
+        // typically less useful anyways.
```

Comment:
I'd be happy to provide a more structured output from the `jsonschema` side to improve the situation here - better handling for `oneOf` was [previously requested](https://github.com/Stranger6667/jsonschema/issues/649), but I didn't work on it yet.

Feel free to tag me on anything related to `jsonschema` crate changes, I am keen to make it more convenient to use :)

The current idea is to store the inner `oneOf` errors and make them accessible through the root `oneOf` error (at least users can have at least some level of introspection)

Reply:
Thanks @Stranger6667, a more structured output here would be amazing! I can file an issue upstream in a bit demonstrating the current behavior (instead of just a `HACK` comment 😅).

(And thank you for making `jsonschema`!)

Reply:
Awesome! A concrete example would be very helpful as a part of an upstream issue :)

And thank you for your kind words

Reply:
Opened https://github.com/Stranger6667/jsonschema/issues/739, let me know if there's any other details that would be helpful!